File: Regresión Lineal/linear_regression_gd.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jan 14 16:41:52 2020

@author: Daniel
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


# Algoritmo de Descenso de Gradiente 
class LinearRegressionGD(object):
    
    def __init__(self, l_rate = 0.1, n_iter =10000):
        self.l_rate = l_rate
        self.n_iter = n_iter               
        
    def fit(self, X, y, theta):
        self.theta = theta
        X_value = X[:,1].reshape(-1, 1)        
        for i in range(0, self.n_iter):
            h = X.dot(self.theta)
            const = self.l_rate*(1/X.shape[0])
            self.theta[0] = self.theta[0]-const*sum(h-y)
            self.theta[1] = self.theta[1]-const*sum((h-y).transpose().dot(X_value))            
        return self.theta
    
    def predict(self, X):
        X_test = X[:, 1]
        predict_value = X_test*self.theta[1]+self.theta[0]
        return predict_value   

# Importar el dataset
dataset = pd.read_csv('Consumo_cerveja.csv')
X = list(dataset.iloc[:365, 2].values)
X = np.asarray([float(a.replace(",",".")) for a in X])
X = X.reshape(-1,1)
X = np.insert(X, 0, 1, axis = 1)
y = dataset.iloc[:365, 6].values

# Seleccionar conjunto de training y test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, 
                                                    random_state = 0)   
   
# Aplicación del modelo
regression = LinearRegressionGD(l_rate = 0.006, n_iter = 100000)
coef = regression.fit(X_train, y_train, np.array([20, 0.6]))
y_predict = regression.predict(X_test)

# Aplicación del modelo con librería de sklearn
from sklearn.linear_model import LinearRegression
regression_py = LinearRegression() 
regression_py.fit(X_train, y_train)
y_predict_py = regression_py.predict(X_test)

# Gráfica de regresión conjunto de test
plt.scatter(X_test[:,1], y_test, color = "red") 
plt.plot(X_test[:,1], y_predict, color = "blue") 
plt.legend(('Descenso de Gradiente',),
           loc='lower right')
plt.title("Consumo de cerveza vs Temperatura Sao Paulo (Conjunto de Test)")
plt.xlabel("Temperatura °C Sao Paulo")
plt.ylabel("Consumo cerveza en Litros")
plt.show()




        

File: Regresión Lineal/linear_regression_gd.R
# Función de Descenso de Gradiente

LinearRegressionGD = function(lrate = 0.1, niter = 10000,
                              X, y, theta){
  const = lrate*(1/length(X))
  for(i in 1:niter){
    h = X*theta[2]+theta[1]
    theta[1] = theta[1]-const*(sum(h-y))
    theta[2] = theta[2]-const*(sum(h-y))*X
  }
  return(theta)
}

# Importar el dataset
dataset = read.csv('Admission_Predict_Ver1.1.csv', sep = ",")
dataset = dataset[1:length(dataset$GRE.Score), c(2,9)]

# Selección conjunto de entrenamiento y test
library(caTools)
set.seed(0)
split = sample.split(dataset$GRE.Score, SplitRatio = 0.75)
training = subset(dataset, split == TRUE)
testing = subset(dataset, split == FALSE)

# Escalado de variables
training = data.frame(scale(training))
testing = data.frame(scale(testing))

# Aplicación del modelo
linearRegression = LinearRegressionGD(lrate = 0.1, niter = 20000,
                                      X = as.numeric(training$GRE.Score),
                                      y = as.numeric(training$Chance.of.Admit),
                                      c(0.1, 1.0))

ypred = testing$GRE.Score*linearRegression[2]+
  linearRegression[1]

# Aplicación del modelo con la librería de R 
regression = lm(formula =  Chance.of.Admit ~ GRE.Score,
               data = training)

# Representación Grafíca
library(ggplot2)

ggplot()+
  geom_point(aes(x=testing$GRE.Score,
                 y=testing$Chance.of.Admit),
             colour = "red")+
  geom_line(aes(x=testing$GRE.Score,
                y=ypred, colour = "blue"),
             alpha = 1,
            size= 0.8)+
  geom_line(aes(x=testing$GRE.Score,
                 y=predict(regression, newdata = testing),
                 colour = "green"), alpha = 1,
             size= 0.8)+
  scale_color_discrete(name = "Modelo", labels = c("Descenso por gradiente", 
                                                   "Regresión Lineal (lm)"))+
  ggtitle("Probabilidad admisión vs GRE Score (Conjunto de Test)")+
  xlab("GRE Score")+
  ylab("Probabilidad admisión a Postgrado")
 

